### PR TITLE
Add vote controller and test.

### PR DESCRIPTION
**Pull request by Murph33**
Make minor changes to views to accomodate voting and place each quote in
a well.

### CODE REVIEW

File: app/controllers/quotes_controller.rb [hunk 1]
```diff
@@ -3,6 +3,7 @@ class QuotesController < ApplicationController
 
   def show
     @quote = Quote.find params[:id]
+    @vote = @quote.votes.find_by(user: current_user)
```
Comment (Senjai) [inline app/controllers/quotes_controller.rb:7]: Does this need to be in the same commit?

File: app/controllers/votes_controller.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
+class VotesController < ApplicationController
+  before_action :authenticate_user!
+
+  def create
```
Comment (Senjai) [inline app/controllers/votes_controller.rb:4]: Each of these actions and their tests can be in their own commits.

File: app/models/quote.rb [hunk 3]
```diff
@@ -4,4 +4,10 @@ class Quote < ActiveRecord::Base
   belongs_to :created_by, class_name: :User, foreign_key: "user_id"
 
   has_many :votes, dependent: :destroy
+
+  def score
```
Comment (Senjai) [inline app/models/quote.rb:8]: Seperate commit here.

File: app/models/quote.rb [hunk 4]
```diff
@@ -4,4 +4,10 @@ class Quote < ActiveRecord::Base
...
+    score = 0
+    votes.each { |vote| score += vote.value }
```
Comment (Senjai) [inline app/models/quote.rb]: `votes.sum(:value)` assuming value is a database column.

File: app/views/quotes/index.html.erb [hunk 5]
```diff
@@ -1,3 +1,6 @@
 <% @quotes.each do |quote| %>
-  <%= render quote %>
+  <div class="well">
```
Comment (Senjai) [inline app/views/quotes/index.html.erb:2]: Each of these views can be in seperate commits with the controller actions.

File: app/views/votes/_voting.html.erb [hunk 6]
```diff
@@ -0,0 +1,14 @@
+<% if Vote.exists?(quote: quote, user: current_user) %>
+  <%= link_to "I hate it", quote_vote_path(quote, vote, { vote: { value: -2 } } ), method: :patch %>
```
Comment (Senjai) [inline app/views/votes/_voting.html.erb]: Magic numbers like -2 and -1 don't really mean anything. We should have this be enum'd in some fashion. `VOTE_VALUES[:hate_it]` or the like. This allows you to write validations on how much someone can vote something down or up, and you only have to change it in one place if you want to tweak these values.

File: app/controllers/quotes_controller.rb (hunk 1 above)
Comment (Murph33) [inline app/controllers/quotes_controller.rb:7]: It would break the quote show page without if I'm remembering right.  Maybe I added more than I should have by including view functionality at the same time?  Was trying to imitate the quote controller commit I'd made before.

**Comment by Senjai** [comment]
Added a few comments primarily focused on breaking this up into smaller commits. If you have to say "as well" in your commit message, it can probably be a separate commit ;)

File: app/views/votes/_voting.html.erb [hunk 7]
```diff
@@ -0,0 +1,14 @@
+<% if Vote.exists?(quote: quote, user: current_user) %>
```
Comment (Senjai) [inline app/views/votes/_voting.html.erb]: This is going to be an N+1 query. For every quote on the page we query to see if the user has voted for this quote or not. I'd like to see this be done better, ideally with a single query for each set of results (or page, if pagination is added at some point).

Comment (Senjai) [inline app/views/votes/_voting.html.erb]: Ignoring perf for the quoteboard. We can ignore this one.

File: app/controllers/votes_controller.rb [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    if @vote.save
+      flash[:notice] = "Voted!"
+    else
+      flash.now[:alert] = "Something went wrong!"
+    end
+    redirect_to quotes_path
+  end
+
+  def update
+    @quote = Quote.find params[:quote_id]
```
Comment (Senjai) [inline app/controllers/votes_controller.rb]: @quote = and @vote = at the beginning of every action can be extracted out to a before_action.

File: app/controllers/application_controller.rb [hunk 9]
```diff
@@ -2,4 +2,14 @@ class ApplicationController < ActionController::Base
   # Prevent CSRF attacks by raising an exception.
   # For APIs, you may want to use :null_session instead.
   protect_from_forgery with: :exception
+
+  helper_method :vote_types
+
+  def vote_types
```
Comment (Murph33) [inline app/controllers/application_controller.rb]: This should be local to the controller that uses it, it does not need to be application wide.

I know later on you switch to the enum approach, we can use I18n at that point as discussed.

~Senjai

**Comment by Murph33** [comment]
This should have been made as a PR to master.  Please see #17
